### ISSUE TITLE
Under the Hood Explanation for other Authors hoping to write mods with APIs

### ISSUE THREAD

**Issue by MalucoMarinero** [andrewgu/ModConfigMenu#29]
Hey there, sorry for clogging up the issues with a request, but I'm at my wits end and you're probably holder of the knowledge I need.

I'm trying to implement the interface approach to exposing a Public API for a mod, for me its going to be for Mission Logic.

I got it all working as a package without that stuff, but now I'm trying to reverse engineer how you set it up and its just not building for me:

https://github.com/GuerrillaWar/XCOMMissionLogic

I think it has to do with the loading order but I don't know well enough how the build scripts do there thing. I noticed that when ModConfigMenu builds it always does the API package first (so the main package always has its depedencies available by the time it compiles), but I'm unsure what or where determines this.

When I pulled down your Mod to test out building it worked first try, so I can only assume I'm doing something wrong but I have no idea what that something is.

Beat my head enough against this tonight so I figured I'm putting out the bat signal.

**Comment by andrewgu** [comment]
Hey @MalucoMarinero happy to help. I wrote it up already over here: https://github.com/andrewgu/ModConfigMenu/blob/master/documentation/sharedcode.md

I'll leave this issue open for now. If that explanation isn't clear enough I'd be happy to chat with you separately about the details.

**Comment by MalucoMarinero** [comment]
Cheers @andrewgu , I actually read through that guide when I was setting it up. The problem I ran into wasn't setting up the dependant mod, but the mod that's sharing it's API. My link goes to the code, essentially the issue is that it always compiles the main package first, in this case XComMissionLogic, before XComMissionLogicAPI. This causes an error because the core package relies on interfaces in the API package. 

I notice that your project doesn't have that compilation error, ModConfigMenuAPI gets built first every time, so when ModConfigMenu compiles it's interfaces are available. Is there something specific you do to cause this to happen? I feel like the repo I linked in the opening post of the issue maps to the structure of yours, I'm not sure what detail I've missed that controls compilation order.

**Comment by andrewgu** [comment]
@MalucoMarinero Interesting observation, I actually wasn't cognizant that this particular issue was implicit in my code.

Maybe try adding an explicit "dependson" to the API interface:

```
class XComGameState_MissionLogic extends XComGameState_BaseObject
    implements(XComGameState_MissionLogic_API)
    dependson(XComGameState_MissionLogic_API, X2TacticalGameRulesetDataStructures);
```

**Comment by MalucoMarinero** [comment]
Yeah okay. If it's something you just never had to solve I guess I'll keep digging and see whether the depends on stuff will help. Cheers.

**Comment by andrewgu** [comment]
Yup, if you don't mind, can you let me know if you find the answer. I'm sure anybody using MCM will also potentially run into this problem.

**Comment by MalucoMarinero** [comment]
So I got to the bottom of it, and like all inexplicable errors, it came from Config. Somehow the config doesn't take for the Engine config

```
[UnrealEd.EditorEngine]
+EditPackages=XComMissionLogicAPI
```

so forcing the matter briefly with:

```
[UnrealEd.EditorEngine]
+EditPackages=XComMissionLogicAPI
+EditPackages=XComMissionLogic
```

actually forced the editor config to refresh, as it seems compile order is governed by the ini's array order.

Unfortunately once I got to this point I realised that I can't use your method of sharing, as my API needs to expose concrete inheritance of my XComGameState, so unfortunately I'm just gonna be stuck with vendoring for now into repos I want to use it on. The interface wrapper solution is a cool one, just one that doesn't quite fit in my scenario. Thanks for the help anyway. Progress made one line of code at a time.

**Comment by andrewgu** [comment]
Hey, thanks for updating! I'm going to add this into the documentation so that future mod authors can get around this problem. Sorry this solution doesn't quite work for you, best of luck with finding an alternative.

**Comment by andrewgu** [comment]
On second thought, I'm keeping this issue open until I've actually added to documentation.
